educator/src/Dscp/Educator/Web/Server.hs

```diff
@@ -1,49 +1,58 @@
+{-# LANGUAGE TypeOperators #-}
+
 -- | Functions to serve Student HTTP API
 
 module Dscp.Educator.Web.Server
        ( serveStudentAPIReal
-       , convertStudentApiHandler
        ) where
 
 import Fmt ((+|), (|+))
 import Loot.Log (logInfo)
-import Servant (Handler, Server, hoistServer, serve, throwError)
+import Servant ((:<|>) (..), (:>), Handler, Server, hoistServer, serve)
 import Servant.Generic (toServant)
 
 import Dscp.Educator.Config (HasEducatorConfig)
-import Dscp.Educator.Launcher (EducatorContext, EducatorRealMode, EducatorWorkMode)
-import Dscp.Educator.Web.Bot.Handlers (addBotHandlers)
+import Dscp.Educator.Launcher (EducatorRealMode, EducatorWorkMode)
+import Dscp.Educator.Web.Bot (addBotHandlers)
+import Dscp.Educator.Web.Educator (EducatorAPI, convertEducatorApiHandler, educatorAPI,
+                                   educatorApiHandlers)
 import Dscp.Educator.Web.Params (EducatorWebParams (..))
-import Dscp.Educator.Web.Student.API (StudentAPI, studentAPI)
-import Dscp.Educator.Web.Student.Error (toServantErr, unexpectedToServantErr)
-import Dscp.Educator.Web.Student.Handlers (studentApiHandlers)
-import Dscp.Launcher.Rio (runRIO)
+import Dscp.Educator.Web.Student (StudentAPI, convertStudentApiHandler, studentAPI,
+                                  studentApiHandlers)
 import Dscp.Web (ServerParams (..), serveWeb)
 
-studentAPIServer
+type EducatorWebAPI = "api" :> (
+    "educator" :> EducatorAPI
+    :<|>
+    "student" :> StudentAPI
+  )
+
+mkEducatorApiServer
+    :: forall ctx m. EducatorWorkMode ctx m
+    => (forall x. m x -> Handler x)
+    -> Server EducatorAPI
+mkEducatorApiServer nat =
+    hoistServer educatorAPI nat (toServant educatorApiHandlers)
+
+mkStudentApiServer
     :: forall ctx m. EducatorWorkMode ctx m
     => (forall x. m x -> Handler x)
     -> Bool
     -> m (Server StudentAPI)
-studentAPIServer nat withBot = do
-    handlers <-
+mkStudentApiServer nat withBot = do
+    studentAndBotHandlers <-
         bool identity (>>= addBotHandlers) withBot
         (pure studentApiHandlers)
```

Comment:
What is going on here?

Reply:
I wanted it to look like a pipeline so that it was easy to build handlers from many pieces, one piece per line